File: Minimum BST Sum Subtree[ 18 April 2023]/Solution.cpp
//    >> C++ CODE <<


class Data{
public:
    int nodeMin;
    int nodeMax;
    int nodeSum;
    int countNode;
    bool isBST;
};

class Solution {
public:
    int mn;
    Data *minSubtree(Node *root, int target) {
        if (root == NULL) {
            Data *p = new Data();
            p->nodeMin = INT_MAX;
            p->nodeMax = INT_MIN;
            p->nodeSum = 0;
            p->countNode = 0;
            p->isBST = true;
            return p;
        }
        Data *left = minSubtree(root->left, target);
        Data *right = minSubtree(root->right, target);
        int nodeSum = left->nodeSum + right->nodeSum + root->data;
        int countNode = left->countNode + right->countNode + 1;
        int nodeMin1 = min(root->data, min(left->nodeMin, right->nodeMin));
        int nodeMax1 = max(root->data, max(left->nodeMax, right->nodeMax));
        bool isBST = root->data > left->nodeMax && root->data < right->nodeMin &&
                        left->isBST && right->isBST;
        if (isBST && nodeSum == target) {
            mn = min(mn, countNode);
        }
        Data *p = new Data();
        p->nodeMin = nodeMin1;
        p->nodeMax = nodeMax1;
        p->nodeSum = nodeSum;
        p->countNode = countNode;
        p->isBST = isBST;
        return p;
    }
    int minSubtreeSumBST(int target, Node *root) {
        mn = INT_MAX;
        minSubtree(root, target);
        if (mn == INT_MAX) mn = -1;
        return mn;
    }
};

File: Minimum BST Sum Subtree[ 18 April 2023]/Solution.java
//  >> JAVA CODE <<


class Solution {
    static int ans;
    static class Data{
        int min,max,sum,count;
        boolean isBST;
        
        Data(boolean isBST,int min,int max,int sum,int count)
        {
            this.isBST=isBST;
            this.min=min;
            this.max=max;
            this.sum=sum;
            this.count=count;
        }
    }
    static Data solve(Node root,int target)
    {
        if(root==null)
        {
            return new Data(true,Integer.MAX_VALUE,Integer.MIN_VALUE,0,0);
        }
        
        Data left=solve(root.left,target);
        Data right=solve(root.right,target);
        
        int totalSum=left.sum+right.sum+root.data;
        int totalCount=left.count+right.count+1;
        int newMin=Math.min(root.data,Math.min(left.min,right.min));
        int newMax=Math.max(root.data,Math.max(left.max,right.max));
        
        boolean isBST= left.isBST && right.isBST && root.data>left.max && root.data<right.min;
        
        if(isBST && totalSum==target) ans=Math.min(ans,totalCount);
        
        return new Data(isBST,newMin,newMax,totalSum,totalCount);
    }
    public static int minSubtreeSumBST(int target, Node root) {
        
        ans=Integer.MAX_VALUE;
        
        solve(root,target);
        
        if(ans==Integer.MAX_VALUE) return -1;
        
        return ans;
    }
}